### PR TITLE
149 View cache

### PR DESCRIPTION
**Pull request by Vahram1995**

### CODE REVIEW

File: src/Libraries/ResourceCache/ViewCache.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+		}
+
+		$cacheFile = $this->getCacheFile($key, $sessionId);
+
+		$data = [
+			'content' => $content,
+			'expires_at' => time() + $ttl,
+		];
+
+		file_put_contents($cacheFile, serialize($data));
```
[the remark this answers is not in the document]
Comment (armanist) [reply src/Libraries/ResourceCache/ViewCache.php:64]: Are we using temp files or regular files?

[the remark this answers is not in the document]
Comment (armanist) [reply src/Libraries/ResourceCache/ViewCache.php:64]: Why we store the `ttl` inside the content?

File: tests/unit/_root/shared/config/config.php [hunk 2]
```diff
@@ -5,5 +5,6 @@
     'lang_default' => 'en',
     'lang_segment' => 1,
     'debug' => 'DEBUG',
-    'test' => 'Testing'
+    'test' => 'Testing',
+	'resource_cache' => true,
```
Comment (armanist) [inline tests/unit/_root/shared/config/config.php:9]: Let's use `cache => true`

**Review by armanist — CHANGES_REQUESTED** [review]
First please look at scrutinizer issues and fix them.

The library uses code PHP functionality, but instead it should use every available functions from framework.

The approach of storing `ttl` into the content is not desirable

serializing the content can lead potential problems

File: src/Router/Route.php [hunk 3]
```diff
@@ -184,6 +196,41 @@ public function middlewares(array $middlewares = []): Route
         return $this;
     }
 
+	public function cacheable(bool $shouldCache, int $ttl = null): Route
```
Comment (Arno2005) [inline src/Router/Route.php:199]: Missing doc block in some methods.

File: tests/unit/Libraries/ResourceCache/ViewCacheTest.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+	public function testGetViewCache()
+	{
+		$this->viewCache->set($this->route, $this->content, $this->sessionId);
+		$viewCache = $this->viewCache->get($this->route, $this->sessionId, $this->ttl);
+
+		$this->assertIsString($viewCache);
+		$this->assertEquals($this->content, $viewCache);
+	}
+
```
Comment (Arno2005) [inline tests/unit/Libraries/ResourceCache/ViewCacheTest.php:49]: The two tests above are the same, if there are no code-wise differences in these tests, maybe remove one of them?

**Review by Arno2005 — CHANGES_REQUESTED** [review]
Add the missing doc blocks and remove duplicate unit tests.

File: src/Mvc/QtView.php [hunk 5]
```diff
@@ -173,7 +179,13 @@ public function render(string $view, array $params = []): ?string
...
         }
 
-        return $this->renderFile($this->layout);
+		$content = $this->renderFile($this->layout);
+
+	    if (route_cache()){
+		    (new ViewCache())->set(route_uri(), $content, session()->getId());
+	    }
+
+        return $content;
```
Comment (armanist) [inline src/Mvc/QtView.php:188]: If route is cacheable and also minifiable, this return still returns normal rendered view

File: src/Router/Route.php [hunk 6]
```diff
@@ -16,6 +16,7 @@
 
 use Quantum\Exceptions\RouteException;
 use Closure;
+use Quantum\Loader\Setup;
```
Comment (armanist) [inline src/Router/Route.php:19]: Keep use statements in order

File: src/Router/Route.php [hunk 7]
```diff
@@ -184,6 +196,41 @@ public function middlewares(array $middlewares = []): Route
         return $this;
     }
 
+	public function cacheable(bool $shouldCache, int $ttl = null): Route
+	{
+		if (!$ttl) {
+			$ttl = config()->get('view_cache.ttl', 300);
+		}
+
+		if (!empty(session()->getId())){
```
Comment (armanist) [inline src/Router/Route.php:205]: if `session()->getId()` is empty the the whole function just returns `$this`, so the above code is unnecessary

```
if (!$ttl) {
			$ttl = config()->get('view_cache.ttl', 300);
		}
```

Comment (armanist) [reply src/Router/Route.php:205]: Use early return instead of nesting

File: src/Router/Route.php [hunk 8]
```diff
@@ -257,4 +304,19 @@ private function assignMiddlewaresToRoute(array &$route, array $middlewares)
...
+		$route['cache'] = [
+			'shouldCache' => $shouldCache,
+			'ttl' => $ttl,
+		];
+	}
+
+	private function canSetCacheToCurrentRoute(): bool
+	{
+		return config()->get('resource_cache') &&
+			!empty(session()->getId()) &&
```
Comment (armanist) [inline src/Router/Route.php:318]: Checking `session()->getId()` might be unnecessary

**Review by armanist — CHANGES_REQUESTED** [review]
There are still several unresolved point, but also check the scrutinizer first

File: src/Libraries/ResourceCache/ViewCache.php [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+		$configCacheDir = config()->get('view_cache.cache_dir', 'cache');
+
+		$this->cacheDir = base_dir() . DS . $configCacheDir . DS . 'views' . DS;
+
+		if ($module = current_module()) {
+			$this->cacheDir = base_dir() . DS . $configCacheDir . DS . 'views' . DS . strtolower($module) . DS;
+		}
+
+		if (!self::$fs->isDirectory($this->cacheDir) && !$fromCommand) {
```
Comment (armanist) [inline src/Libraries/ResourceCache/ViewCache.php:51]: Please check the need of `$fromCommand`

File: src/Libraries/ResourceCache/ViewCache.php [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+	/**
+	 * @param string $key
+	 * @param string $sessionId
+	 * @param int $ttl
+	 * @return bool
+	 */
+	public static function exists(string $key, string $sessionId, int $ttl): bool
+	{
+		$cacheFile = (new self())->getCacheFile($key, $sessionId);
```
Comment (armanist) [inline src/Libraries/ResourceCache/ViewCache.php:117]: Why the method called via `new self()` vs `$this`

File: src/Libraries/ResourceCache/ViewCache.php [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+	 * @throws ConfigException
+	 * @throws DiException
+	 * @throws ReflectionException
+	 */
+	public function __construct(bool $fromCommand = false)
+	{
+		self::$fs = Di::get(FileSystem::class);
+
+		if (!config()->has('view_cache') && !$fromCommand) {
+			throw ConfigException::configCollision('view_cache');
```
Comment (armanist) [inline src/Libraries/ResourceCache/ViewCache.php:40]: `configCollision()` is not correct throw in this case

File: src/Mvc/MvcManager.php [hunk 12]
```diff
@@ -34,113 +39,127 @@
...
+		if (current_middlewares()) {
+			list($request, $response) = (new MiddlewareManager())->applyMiddlewares($request, $response);
+		}
+
+		$callback = route_callback();
+		$cacheSettings = route_cache_settings();
+
+		if ($cacheSettings &&
+			session()->getId() &&
+			ViewCache::exists(route_uri(), session()->getId(), $cacheSettings['ttl'])
```
Comment (armanist) [inline src/Mvc/MvcManager.php:68]: `ViewCache::exists` creates instance inside the method, and the next instance is created few lines bellow.
It leads to apply singleton approach

File: src/Mvc/MvcManager.php [hunk 13]
```diff
@@ -34,113 +39,127 @@
...
+		){
+			call_user_func_array(function () use (&$response, $cacheSettings) {
+				$content = (new ViewCache())->get(route_uri(), session()->getId(), $cacheSettings['ttl']);
```
Comment (armanist) [inline src/Mvc/MvcManager.php:71]: Once the `ViewCache` implements singleton the access to the class can be unified

File: src/Mvc/QtView.php [hunk 14]
```diff
@@ -173,7 +179,15 @@ public function render(string $view, array $params = []): ?string
             AssetManager::getInstance()->register($this->assets);
         }
 
-        return $this->renderFile($this->layout);
+		$content = $this->renderFile($this->layout);
+
+	    if (($cacheSettings = route_cache_settings()) && session()->getId()){
+		    $content = (new ViewCache())
```
Comment (armanist) [inline src/Mvc/QtView.php:185]: `ViewCache` will be access via singleton

File: src/Libraries/ResourceCache/ViewCache.php [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+namespace Quantum\Libraries\ResourceCache;
+
+use Quantum\Libraries\Storage\FileSystem;
+use Quantum\Exceptions\ConfigException;
+use Quantum\Exceptions\DiException;
+use ReflectionException;
+use voku\helper\HtmlMin;
+use Quantum\Di\Di;
+
+class ViewCache
```
Comment (armanist) [inline src/Libraries/ResourceCache/ViewCache.php:12]: Implement singleton pattern

**Review by armanist — CHANGES_REQUESTED** [review]

File: src/Libraries/ResourceCache/ViewCache.php [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+		if (time() > ($this->fs->lastModified($cacheFile) + $ttl)) {
+			$this->fs->remove($cacheFile);
+			return false;
+		}
+
+		return true;
+	}
+
+	private static function getCacheDir(): string
```
Comment (armanist) [inline src/Libraries/ResourceCache/ViewCache.php:138]: Use non static functions

File: src/Libraries/ResourceCache/ViewCache.php [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+		}
+	}
+
+	/**
+	 * @param string $key
+	 * @param string $sessionId
+	 * @param int $ttl
+	 * @return bool
+	 */
+	public function exists(string $key, string $sessionId, int $ttl): bool
```
Comment (armanist) [inline src/Libraries/ResourceCache/ViewCache.php:122]: Method `exists()` doesn't follow single responsibility principle, it does check as well as remove

**Review by armanist — CHANGES_REQUESTED** [review]

File: src/Libraries/ResourceCache/ViewCache.php [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+	/**
+	 * @param string $key
+	 * @param string $sessionId
+	 * @param int $ttl
+	 * @return mixed|null
+	 */
+	public function get(string $key, string $sessionId, int $ttl): ?string
+	{
+		$cacheFile = $this->getCacheFile($key, $sessionId);
+		if (!$this->fs->exists($cacheFile)) {
```
Comment (armanist) [inline src/Libraries/ResourceCache/ViewCache.php:97]: It should use `$this->exists()`

**Review by armanist — CHANGES_REQUESTED** [review]

File: tests/unit/Router/RouteTest.php [hunk 19]
```diff
@@ -22,7 +24,9 @@ public function setUp(): void
             ]
         ];
 
-        $this->route = new Route($module);
+	    config()->set('session.driver', 'file');
+	    Mockery::mock('Quantum\Libraries\ResourceCache\ViewCache');
```
Comment (armanist) [inline tests/unit/Router/RouteTest.php:28]: Why we are mocking `ViewCache` here?

File: src/Router/Router.php [hunk 20]
```diff
@@ -113,6 +114,10 @@ public function findRoute()
 
         $currentRoute = $this->currentRoute();
 
+	    if (is_array($currentRoute) && key_exists('shouldCache', $currentRoute)){
```
Comment (armanist) [inline src/Router/Router.php:117]: This section should be placed after this condition, and there wouldn't need to check it's array or not

```
if (!$currentRoute) {
    throw RouteException::incorrectMethod($this->request->getMethod());
}
```

**Review by armanist — CHANGES_REQUESTED** [review]
The latest changes introduced some inconsistencies.. 

The `ttl` property of `ViewCache` is set from `Route`  class while `isEnabled` is set from `Router` class.

We need to keep consistency on similar operations applied on same object.

It's obvious now that these both things (`ttl` and `isEnabled`) should be set from `Router` class after the current route is known. Therefore there would no need of passing `ViewCache` instance to `Route` class anymore.

**Review by armanist — APPROVED** [review]

**Review by andrey-smaelov — APPROVED** [review]

**Review by Arno2005 — APPROVED** [review]

**Pull request merged**